File: Leetcode/03_Hash-Table/019_0454_4sum-ii.go
// https://leetcode.cn/problems/4sum-ii/
// m

package main

func fourSumCount(nums1, nums2, nums3, nums4 []int) (count int) {
	sum12Count := map[int]int{}

	// 计算 nums1+nums2 的所有可能和
	for _, n1 := range nums1 {
		for _, n2 := range nums2 {
			sum12Count[n1+n2]++
		}
	}

	// 计算 -(nums3+nums4) 的出现次数
	for _, n3 := range nums3 {
		for _, n4 := range nums4 {
			count += sum12Count[-n3-n4] // 不存在的key会返回0
		}
	}
	return
}

File: Leetcode/03_Hash-Table/019_0454_4sum-ii.py
# https://leetcode.cn/problems/4sum-ii/
# m

class Solution(object):
    '''
    https://leetcode.cn/problems/4sum-ii/solutions/65894/chao-ji-rong-yi-li-jie-de-fang-fa-si-shu-xiang-jia/
    '''
    def fourSumCount(self, nums1, nums2, nums3, nums4):
        dct = dict()
        for n1 in nums1:
            for n2 in nums2:
                dct[n1+n2] = dct.get(n1+n2, 0) + 1

        res = 0
        for n3 in nums3:
            for n4 in nums4:
                target = -n3 - n4
                res += dct.get(target, 0)
        
        return res